elements/lisk-p2p/src/peer_pool.ts

```diff
@@ -603,53 +607,58 @@ export class PeerPool extends EventEmitter {
 		);
 
 		// Cannot predict which netgroups will be protected
-		const filteredPeersByNetgroup = this._peerPoolConfig.netgroupProtectionRatio
+		const protectedPeersByNetgroup = this._peerPoolConfig
+			.netgroupProtectionRatio
 			? filterPeersByCategory(peers, {
 					category: PROTECTION_CATEGORY.NET_GROUP,
 					percentage: this._peerPoolConfig.netgroupProtectionRatio,
 					asc: true,
 			  })
-			: peers;
-		if (filteredPeersByNetgroup.length <= 1) {
-			return filteredPeersByNetgroup;
-		}
+			: [];
 
 		// Cannot manipulate without physically moving nodes closer to the target.
-		const filteredPeersByLatency = this._peerPoolConfig.latencyProtectionRatio
+		const protectedPeersByLatency = this._peerPoolConfig.latencyProtectionRatio
 			? filterPeersByCategory(peers, {
 					category: PROTECTION_CATEGORY.LATENCY,
 					percentage: this._peerPoolConfig.latencyProtectionRatio,
-					asc: true,
+					asc: false,
 			  })
-			: filteredPeersByNetgroup;
-		if (filteredPeersByLatency.length <= 1) {
-			return filteredPeersByLatency;
-		}
+			: [];
 
 		// Cannot manipulate this metric without performing useful work.
-		const filteredPeersByResponseRate = this._peerPoolConfig
+		const protectedPeersByResponseRate = this._peerPoolConfig
 			.productivityProtectionRatio
-			? filterPeersByCategory(filteredPeersByLatency, {
+			? filterPeersByCategory(peers, {
 					category: PROTECTION_CATEGORY.RESPONSE_RATE,
 					percentage: this._peerPoolConfig.productivityProtectionRatio,
-					asc: false,
+					asc: true,
 			  })
-			: filteredPeersByLatency;
-		if (filteredPeersByResponseRate.length <= 1) {
-			return filteredPeersByResponseRate;
-		}
+			: [];
+
+		const uniqueProtectedPeers = [
+			...new Set([
+				...protectedPeersByNetgroup,
+				...protectedPeersByLatency,
+				...protectedPeersByResponseRate,
+			]),
```

Comment:
Why put the Set into an array? The Set is much faster for lookup especially if there are a lot of peers.

In the worst case, `Array.includes()` iterates over every element whereas `Set.has()` hashes directly to the key. Also, we should store the peer ids instead of the whole `Peer` instances or else it can lead to mismatched references if one object is a clone (which is a common thing that pure functions do); it's safer to always use primitive types for lookups.